File: py/mars.py
from pprint import pformat

from const import chars
from grid import Grid
from rover import Rover


class Mars:
    def __init__(self, rows=10, obstacles=5):
        self.rows = rows
        self.obstacles = obstacles
        self.mars = None
        self.grid = None
        self.rovers = []
        self._create()

    def _create(self):
        self.mars = Grid(rows=self.rows, obstacles=self.obstacles)
        self.grid = self.mars.grid

        print("Welcome to Mars!")

        count = 0

        for x in range(len(self.grid)):
            count += self.grid[x].count("O")

        print(f"• Total obstacles: {count} \n")

    def add_rover(self, name, x=0, y=0, direction="N"):
        name = name.capitalize()
        direction = direction.upper()

        if self.grid[y][x]:
            raise Exception(f"Error: Can't place {name} at {x}, {y}")

        print(f"{name} is landing on Mars!")

        rover = Rover(name, x, y, direction, self.mars)
        self.rovers.append(rover)
        self.grid[y][x] = chars["rover"]

        return rover

    def print_grid(self):
        print(pformat(self.grid, indent=2), "\n")

File: js/node/mars.js
const Rover = require('./rover');
const Grid = require('./grid');
const {obsChar, roverChar, defaults} = require('./const');

module.exports = class Mars {
  constructor(rows = defaults.rows, obstacles = defaults.obstacles) {
    this.rows = rows;
    this.obstacles = obstacles;
    this.mars = null;
    this.grid = null;
    this.rovers = [];
    this.create();
  }

  create() {
    let count = 0;

    this.mars = new Grid(this.rows, this.obstacles);
    this.grid = this.mars.grid;

    console.log('Welcome to Mars!');

    this.grid.forEach((row) => {
      count += row.filter((e) => e === obsChar).length;
    });

    console.log(`• Total obstacles: ${count} \n`);
  }

  addRover(
      name, x = defaults.x, y = defaults.y, direction = defaults.direction,
  ) {
    name = name.charAt(0).toUpperCase() + name.slice(1);
    direction = direction.toUpperCase();

    if (this.grid[y][x]) {
      throw new Error(`Error: Can't place ${name} at ${x}, ${y}`);
    }

    console.log(`${name} is landing on Mars!`);

    const rover = new Rover(name, x, y, direction, this.mars);

    this.rovers.push(rover);
    this.grid[y][x] = roverChar;

    return rover;
  }

  printGrid() {
    console.log(this.grid, '\n');
  }
};
